cpp/modmesh/view/RCameraController.cpp

```diff
@@ -0,0 +1,343 @@
+
+    , m_shiftButtonAction(new Qt3DInput::QAction)
+    , m_altButtonAction(new Qt3DInput::QAction)
+
+    , m_leftMouseButtonInput(new Qt3DInput::QActionInput)
+    , m_middleMouseButtonInput(new Qt3DInput::QActionInput)
+    , m_rightMouseButtonInput(new Qt3DInput::QActionInput)
+
+    , m_shiftButtonInput(new Qt3DInput::QActionInput)
+    , m_altButtonInput(new Qt3DInput::QActionInput)
+
+    , m_mouseRxInput(new Qt3DInput::QAnalogAxisInput)
+    , m_mouseRyInput(new Qt3DInput::QAnalogAxisInput)
+    , m_mouseTzXInput(new Qt3DInput::QAnalogAxisInput)
+    , m_mouseTzYInput(new Qt3DInput::QAnalogAxisInput)
+
+    , m_keyboardTxPosInput(new Qt3DInput::QButtonAxisInput)
+    , m_keyboardTyPosInput(new Qt3DInput::QButtonAxisInput)
+    , m_keyboardTzPosInput(new Qt3DInput::QButtonAxisInput)
+    , m_keyboardTxNegInput(new Qt3DInput::QButtonAxisInput)
+    , m_keyboardTyNegInput(new Qt3DInput::QButtonAxisInput)
+    , m_keyboardTzNegInput(new Qt3DInput::QButtonAxisInput)
+{
+    init();
+
+    connect(m_frameAction, &Qt3DLogic::QFrameAction::triggered,
+    this, [this] (const float dt) {
+        Qt3DExtras::QAbstractCameraController::InputState state{};
+
+        state.rxAxisValue = m_rxAxis->value();
+        state.ryAxisValue = m_ryAxis->value();
+        state.txAxisValue = m_txAxis->value();
+        state.tyAxisValue = m_tyAxis->value();
+        state.tzAxisValue = m_tzAxis->value();
+
+        state.leftMouseButtonActive = m_leftMouseButtonAction->isActive();
+        state.middleMouseButtonActive = m_middleMouseButtonAction->isActive();
+        state.rightMouseButtonActive = m_rightMouseButtonAction->isActive();
+
+        state.altKeyActive = m_altButtonAction->isActive();
+        state.shiftKeyActive = m_shiftButtonAction->isActive();
+
+        m_callback(state, dt);
+    });
+}
+
+void RCameraInputListener::init()
+{
+    // left mouse button
+    m_leftMouseButtonInput->setButtons(QList<int> { Qt::LeftButton });
+    m_leftMouseButtonInput->setSourceDevice(m_mouseDevice);
+    m_leftMouseButtonAction->addInput(m_leftMouseButtonInput);
+
+    // middle mouse button
+    m_middleMouseButtonInput->setButtons(QList<int> { Qt::MiddleButton });
+    m_middleMouseButtonInput->setSourceDevice(m_mouseDevice);
+    m_middleMouseButtonAction->addInput(m_middleMouseButtonInput);
+
+    // right mouse button
+    m_rightMouseButtonInput->setButtons(QList<int> { Qt::RightButton });
+    m_rightMouseButtonInput->setSourceDevice(m_mouseDevice);
+    m_rightMouseButtonAction->addInput(m_rightMouseButtonInput);
+
+    // shift button
+    m_shiftButtonInput->setButtons(QList<int> { Qt::Key_Shift });
+    m_shiftButtonInput->setSourceDevice(m_keyboardDevice);
+    m_shiftButtonAction->addInput(m_shiftButtonInput);
+
+    // alt button
+    m_altButtonInput->setButtons(QList<int> { Qt::Key_Alt });
+    m_altButtonInput->setSourceDevice(m_keyboardDevice);
+    m_altButtonAction->addInput(m_altButtonInput);
+
+    //// Axes
+
+    // mouse x rotation
```

Comment:
mouse x/y rotation is listening for mouse's movement, which can be only in two dimensions.

mouse x/z translation is listening for mouse's wheels. Afaik, there are mouses with trackballs, which can rotate in two dimensions. 

Actually, we can remove mouse x translation, since such mouses aren't common. I decided to keep it, because qt had it, and I wanted as little changes to camera's behavior as possible.

Reply:
This is worth mentioning in the code remark.  I agree keeping support for trackballs is good.  Professional CAD may even use more moving modes for pointing devices.